### ISSUE TITLE
Reduce duplication of format and check scripts

### ISSUE THREAD

**Issue by nickmccurdy** [prettier/prettier#4129]
Most projects I've seen using Prettier have local configs to support editor plugins and build scripts using `--write` and `--list-different` for local formatting and CI checks respectively. For example, here's what Redux uses in its `package.json` scripts:

```json
"format": "prettier --write \"{src,test}/**/*.js\"",
"format:check": "prettier --list-different \"{src,test}/**/*.js\"",
```

I see this duplication of file arguments often because prettier configs don't have file lists, and I'd like to find a way to make this use case cleaner.

## Options
1. Adjust the `--write` flag to behave like `--list-different` when the `CI` environment variable is true. Most CI environments set this variable, and it's already used by various tools to prevent writing to files in CI environments (for example, Jest snapshots aren't written on CI environments by default). Then pairs of scripts like these could simply be replaced with a single `--write` script. This could technically break existing CI builds that accidentally reformatted code.
2. Same as 1 but use a new flag like `--auto` to keep back compatible behavior in CI.
3. Provide a separate config file allowing non-formatting options so you can define the file list once.
4. Allow the `--list-different` and `--write` flags to disable each other (enabling whichever one was specific last) to allow scripts to be composed by appending overriding arguments.

**Comment by evilebottnawi** [comment]
@nickmccurdy who don't use?
```
"format": "npm run format:check -- --write",
"format:check": "prettier --list-different \"{src,test}/**/*.js\"",
```

**Comment by nickmccurdy** [comment]
Doesn't that still fail when files are unformatted? And either way, I still like the idea of only having to add a single command to `package.json` scripts because this use case is so common. Prettier should act similarly across projects, right now users have to memorize the setups their packages are using and the scripts often differ significantly.

**Comment by duailibe** [comment]
Yeah I'm currently thinking an API that makes more sense is to remove `--list-different` and make the default be that. Users can then add to `package.json`:
```
"prettier": "prettier \"{src,test}/**/*.js\"",
```

And just `npm run prettier` to list files that changed and `npm run prettier -- --write` to write files. The `--list-different` is really weird.

(I also think this is being discussed already in some issue but I can't remember)

**Comment by nickmccurdy** [comment]
Alternatively, I would prefer to have `--write` as the default normally and `--list-different` as the default in CI when paths are given without either flag. If neither flag nor any paths are given, it uses stdout as with the current default behavior to prevent integrations from breaking.

**Comment by j-f1** [comment]
I don’t think it’s a good idea to rewrite people’s code without them telling you to do so.

**Comment by duailibe** [comment]
Even though most users will just always add `--write` to the command because they never run Prettier without it, we still want new users to have a "safe" experience when trying Prettier.

**Comment by nickmccurdy** [comment]
> Even though most users will just always add --write to the command because they never run Prettier without it, we still want new users to have a "safe" experience when trying Prettier.

In that case, I think we can achieve this by loading defaults onto Prettier's `--write` command (so you have to provide it explicitly but `--list-different` would be called implicitly on CI). Then without either flag, Prettier would remain back compatible for tooling integrations and new users.

**Comment by suchipi** [comment]
I don't think making the default behavior different in CI is a good idea. It will confuse users who are trying to set up their CI when it behaves differently than on their machine.

**Comment by nickmccurdy** [comment]
In my opinion changing the behavior on CI would lead to less confusion, since `--write` isn't well suited for CI environments. Most CI services will not commit code after it's formatted (so it would be preferred to use `--list-different`, which we could specify implicitly, to show what you be changed locally). Note that existing tools like Jest snapshots already do this. Also I think disabling all file writing in CI is an important safety consideration, since some build pipelines will send formatted code to a production environment after the build even if it's not committed. This could theoretically cause Prettier to break production deploys when users are not aware of formatting occurring in CI, and I think failing fast like `--list-different` would be both safer and more useful.

Alternatively if we don't want to affect back compatibility at all, my suggestions could still be applied to a new flag. I like the name `--auto`, and it would behave exactly like my previous suggestions for `--write`.

**Comment by rhys-vdw** [comment]
Surely the best solution would be to have a way to specify checked paths in the `.prettierconfig`?

At the moment we have this:

```
    "format": "prettier --write './{{app/assets,config,stories}/**/*.{ts,tsx,js,jsx,scss,css,md},*.md}'",
```

And to be honest I'd like to support more but jamming all these things into a single blob is getting unmanageable.

**Comment by nickmccurdy** [comment]
Prettier now supports `.` which makes CLI configuration much simpler. While having a separate script is simpler now, it's still possible to forget that it works differently in CI and only have one of the two scripts. Perhaps it would still be useful to have a new flag like `--auto` which checks in CI and formats otherwise, so users could opt in to a single script.

**Comment by jasonkarns** [comment]
The goal here is for a single command to both --write but also support running on CI, right? So that a single `npm run lint` task is sufficient.

All that's necessary for CI is for `--write` to exit with the proper status code. If files were modified, exit with an error code. (The --write output already has the list of files modified.) Why not just have `--write` and `--check` flags supported in combination such that the exit status is reported correctly and CI can respond in kind? 

I can't recall if it's standard or prettier-standard, but one of the two already works this way. It reports exit status 0 when no files needed formatted; 1 if files were formatted.

Of course, instead of overloading the `--check` flag, it could also be handled by an `--exit-code` flag like, `git-status`.

**Comment by nickmccurdy** [comment]
I know this issue if old, but if anyone's interested I can try to implement the `--auto` flag I suggested above. Unlike `--write` it would be able to save time by not writing to files on CI, since CI environments typically don't cache or publish the original source code anyway.

**Comment by what1s1ove** [comment]
> I know this issue if old, but if anyone's interested I can try to implement the `--auto` flag I suggested above. Unlike `--write` it would be able to save time by not writing to files on CI, since CI environments typically don't cache or publish the original source code anyway.

It would be grate 🙂

**Comment by varthc** [comment]
I was googling and found this open issue. Any update or solution?
I tried something similar to what @alexander-akait proposed above (https://github.com/prettier/prettier/issues/4129#issuecomment-372121536), however I'm still not covering all my scenarios.
I want to run prettier manually (command line), with lint staged and in the pipeline, however I'm trying to minimize the configs and scripts.
With those 2 scripts, I need to duplicate the config in lint staged or create another script for it. 
Has anyone faced this and solved it? Thanks
